src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -226,25 +285,148 @@ bool DeepChangeChecker::check_row(Table const& table, ObjKeyType key, size_t dep
+                                                               bool all_callbacks_filtered)
+    : DeepChangeChecker(info, root_table, related_tables, key_path_array, all_callbacks_filtered)
+{
+}
+
+bool CollectionKeyPathChangeChecker::operator()(ObjKeyType object_key)
+{
+    std::vector<int64_t> changed_columns = {};
+
+    for (auto& key_path : m_key_path_array) {
+        find_changed_columns(changed_columns, key_path, 0, m_root_table, object_key);
+    }
+
+    return changed_columns.size() > 0;
+}
+
+void CollectionKeyPathChangeChecker::find_changed_columns(std::vector<int64_t>& changed_columns,
+                                                          const KeyPath& key_path, size_t depth, const Table& table,
+                                                          const ObjKeyType& object_key_value)
+{
+
+    if (depth >= key_path.size()) {
+        // We've reached the end of the key path.
+
+        // For the special case of having a backlink at the end of a key path we need to check this level too.
+        // Modifications to a backlink are found via the insertions on the origin table (which we are in right
+        // now).
+        auto last_key_path_element = key_path[key_path.size() - 1];
+        auto last_column_key = last_key_path_element.second;
+        if (last_column_key.get_type() == col_type_BackLink) {
+            auto iterator = m_info.tables.find(table.get_key().value);
+            if (iterator != m_info.tables.end() && !iterator->second.insertions_empty()) {
+                auto root_column_key = key_path[0].second;
+                changed_columns.push_back(root_column_key.value);
+            }
+        }
+
+        return;
+    }
+
+    auto [table_key, column_key] = key_path.at(depth);
+
+    // Check for a change on the current depth level.
+    auto iterator = m_info.tables.find(table_key.value);
+    if (iterator != m_info.tables.end() && (iterator->second.modifications_contains(object_key_value, {column_key}) ||
+                                            iterator->second.insertions_contains(object_key_value))) {
+        // If an object linked to the root object was changed we only mark the
+        // property of the root objects as changed.
+        // This is also the reason why we can return right after doing so because we would only mark the same root
+        // property again in case we find another change.
+        auto root_column_key = key_path[0].second;
+        changed_columns.push_back(root_column_key.value);
+        return;
+    }
+
+    // Advance one level deeper into the key path.
+    if (column_key.is_list()) {
+        auto target_table = table.get_link_target(column_key);
+        auto object = table.get_object(ObjKey(object_key_value));
+        auto lvr = object.get_linklist(ColKey(column_key));
+        auto sz = lvr.size();
+        for (size_t i = 0; i < sz; i++) {
+            find_changed_columns(changed_columns, key_path, depth + 1, *target_table, lvr.get(i).value);
```

Comment:
In these loops, can't we return early as soon as one change is found?

Reply:
We need to find all changed columns for objects so we need to run though the whole loop to not miss out on one of those indices.